client/my-sites/themes/index.web.js

```diff
@@ -41,6 +41,8 @@ export default function( router ) {
 				makeLayout
 			);
 		} else {
+			router( '/themes/upload*', '/themes' );
```

Comment:
Would be nice if the redirect was a bit more explicit, but I see that relying on `router.redirect()` to be translated to `page.redirect()` is risky, too (implementation detail). Let's keep it and maybe add a little comment (upload not possible when logged-out, hence redirect).